horovod/common/ops/adasum_cuda_ring_operations.cc

```diff
@@ -0,0 +1,198 @@
+// Copyright 2019 Microsoft. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include "adasum_cuda_ring_operations.h"
+
+namespace horovod {
+namespace common {
+
+AdasumCudaRingAllreduceOp::AdasumCudaRingAllreduceOp(MPIContext* mpi_context, GPUContext* gpu_context, HorovodGlobalState* global_state)
+    : AdasumMPI(mpi_context, global_state), GPUAllreduce(gpu_context, global_state){
+}
+
+AdasumCudaRingAllreduceOp::~AdasumCudaRingAllreduceOp() {
+}
+
+Status AdasumCudaRingAllreduceOp::Execute(std::vector<TensorTableEntry>& entries, const Response& response) {
+  if(entries.empty()) {
+    return Status::OK();
+  }
+  gpu_op_context_.InitGPU(entries);
+  if (!reduction_comms_initialized) {
+    InitializeVHDDReductionComms();
+  }
+  return RingHierarchical(entries, response);
+}
+
+Status AdasumCudaRingAllreduceOp::RingHierarchical(std::vector<TensorTableEntry>& entries,
+                        const Response& response) {
+
+  int num_reductions = entries.size();
+
+  // Because Adasum ring does not currently support fusion, use the fusion buffer for the temp buffer
+  auto& first_entry = entries[0];
+  auto global_buffer = global_state_->fusion_buffer.GetBuffer(
```

Comment:
This is kinda dangerous, what if users set fusion buffer to 0. I dont think we will need fusion for rings, it will make the pipeline inefficient. Maybe just allocate our own buffer here.

Reply:
This is true, I thought about this as well. However, the fusion buffer is already the right size and allocated on the right device, so not using it seemed kind of wasteful of memory to me, especially since it's not being used for anything else. I could add a warning saying that this requires fusion buffer to be set to nonzero, or do you think it's better from a user perspective to just allocate my own buffer here even if it uses some extra memory?

Reply:
I think it's better to have separate buffer. There might be other operations running on the same device that use this fusion buffer, we would need to make sure for the ring reduction we don't affect fusion for other operations and vice versa.

Reply:
Agreed, working on changing this.